File: python/examples/postgres_psql.py
import argparse
import asyncio
import shutil
import subprocess

from wasmer_sdk import Wasmer

PORT = 5432


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Run PostgreSQL under Wasmer and query it with native psql"
    )
    parser.add_argument(
        "--psql",
        default=shutil.which("psql"),
        help="path to the native psql client (defaults to PATH)",
    )
    return parser.parse_args()


async def wait_for_line(lines, marker: str) -> None:
    async for line in lines:
        print(line)
        if marker in line:
            return
    raise RuntimeError(f"PostgreSQL exited before emitting {marker!r}")


def query(psql: str, port: int) -> subprocess.CompletedProcess[str]:
    uri = (
        f"postgresql://postgres@127.0.0.1:{port}/"
        "postgres?sslmode=disable"
    )
    return subprocess.run(
        [
            psql,
            uri,
            "-X",
            "-v",
            "ON_ERROR_STOP=1",
            "-At",
            "-c",
            "select version(), 40 + 2 as answer;",
        ],
        capture_output=True,
        text=True,
        timeout=10,
        check=False,
    )


async def main() -> None:
    args = arguments()
    if not args.psql:
        raise SystemExit("psql was not found; install it or pass --psql")

    async with Wasmer(output_bytes=256 * 1024) as client:
        pglite = await client.packages.load("wasmer/pglite@0.1.0")
        sandbox = await client.sandboxes.create(
            packages=[pglite],
            network="host",
        )
        async with sandbox:
            process = await sandbox.command(pglite).spawn(
                stdout="capture",
                stderr="pipe",
                output_bytes=256 * 1024,
            )
            assert process.stderr is not None

            try:
                await asyncio.wait_for(
                    wait_for_line(
                        process.stderr.lines(),
                        f"OLIPHAUNT_WASIX_SOCKET_READY {PORT}",
                    ),
                    timeout=30,
                )
                result = await asyncio.to_thread(query, args.psql, PORT)
                if result.returncode != 0:
                    raise RuntimeError(
                        f"psql exited with {result.returncode}\n"
                        f"{result.stderr}"
                    )
                print(result.stdout.strip())
                await asyncio.wait_for(
                    process.wait(check=True),
                    timeout=10,
                )
            finally:
                await process.kill()
                await process.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: js/examples/postgres_psql.mjs
import assert from "node:assert/strict";
import { execFile } from "node:child_process";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";

import { Wasmer } from "@wasmer/sdk2/node";

const execFileAsync = promisify(execFile);
const port = 5432;
const psql = process.env.PSQL ?? process.argv[2] ?? "psql";
const query = fileURLToPath(
  new URL("../../fixtures/postgres/query.sql", import.meta.url),
);

const client = new Wasmer({ outputBytes: 256 * 1024 });
const sandbox = await client.sandboxes.create({
  packages: ["wasmer/pglite@0.1.0"],
  network: { mode: "host" },
});
const postgres = await sandbox
  .command("pglite")
  .spawn({ stdout: "capture", stderr: "pipe" });

try {
  assert(postgres.stderr);
  await waitForLine(
    postgres.stderr.lines(),
    `OLIPHAUNT_WASIX_SOCKET_READY ${port}`,
    30_000,
  );

  const uri =
    `postgresql://postgres@127.0.0.1:${port}/postgres?sslmode=disable`;
  const result = await execFileAsync(
    psql,
    [uri, "-X", "-v", "ON_ERROR_STOP=1", "-At", "-f", query],
    { timeout: 10_000 },
  );
  console.log(`Connected native psql to PostgreSQL running in Wasmer`);
  console.log(result.stdout.trim());

  await postgres.wait({ check: true });
} catch (error) {
  await postgres.kill();
  const output = await postgres.wait();
  throw new Error(`${error}\nPostgreSQL stderr:\n${output.stderr.text()}`);
} finally {
  await sandbox.close();
  await client.close();
}

async function waitForLine(lines, marker, timeoutMs) {
  let timeout;
  try {
    await Promise.race([
      (async () => {
        for await (const line of lines) {
          if (line.includes(marker)) return;
        }
        throw new Error(`process exited before emitting ${marker}`);
      })(),
      new Promise((_, reject) => {
        timeout = setTimeout(
          () => reject(new Error(`timed out waiting for ${marker}`)),
          timeoutMs,
        );
      }),
    ]);
  } finally {
    clearTimeout(timeout);
  }
}
